assignment-client/src/audio/AudioMixerSlavePool.cpp

```diff
@@ -53,15 +53,16 @@ void AudioMixerSlaveThread::wait() {
 }
 
 void AudioMixerSlaveThread::notify(bool stopping) {
-    {
-        Lock lock(_pool._mutex);
-        assert(_pool._numFinished < _pool._numThreads);
-        ++_pool._numFinished;
-        if (stopping) {
-            ++_pool._numStopped;
-        }
+    Lock poolLock(_pool._poolMutex);
+    assert(_pool._numFinished < _pool._numThreads);
+    ++_pool._numFinished;
+    if (stopping) {
+        ++_pool._numStopped;
+    }
+
+    if(_pool._numFinished == _pool._numThreads) {
+        _pool._poolCondition.notify_one();
```

Comment:
I’m not familiar with this code so I’m still parsing all of this, but are we sure this change is safe?  If we kick off a bunch of workers, when one finishes, shouldn’t it immediately notify the pool to start another one?  with this change, aren’t we waiting for them all to finish before starting new ones?

Reply:
if you look at the _poolCondition.wait calls, they all use a precondition of "_pool._numFinished == _pool._numThreads".  The idea is the pool wakes up *all* the current slaves, they do their work, then the pool wakes back up when they're finished (and go back to sleep again).  Without this "if" statement the pool wait will have a spurious wake every time each of the slaves finishes its work (but won't leave the wait() call).

Reply:
that seems weird…shouldn’t we only be waking up workers as needed like in a normal thread pool?  why do we do them all at once?  isn’t it possible some of them won’t have anything to do?

Reply:
The goal is to consume the contents of the queue as quickly as possible, if the queue is empty then the (worker thread?) flags as "finished" without doing anything.

I'm unsure why we aren't using a threadpool on this (other than "spawn 100 tasks" might be a bit weird?).  The only thing I can think of is that this is optimized the heck out to be as fast as possible with as little task setup/shutdown as possible.

Reply:
ok thank you that makes sense!